### PR TITLE
✨ Handle arbitrary metrics order and add testing

### PR DESCRIPTION
**Pull request by superbuggy**
Changes:
- Handles any ordering of CVSS metrics (is this desirable?)
- Defines test suite
- Defines continuous integration workflow

### CODE REVIEW

File: cvss40.js [hunk 1]
```diff
@@ -437,13 +446,18 @@ class Vector {
      */
     updateMetricsFromVectorString(vector) {
         if (!vector) {
-            throw new Error("The vector string cannot be null, undefined, or empty.");
+            throw new Error(`The vector string cannot be null, undefined, or empty. in ${vector}`);
```
Comment (skontar) [inline cvss40.js]: Full stop in the middle of the sentence.

File: cvss40.js [hunk 2]
```diff
@@ -437,13 +446,18 @@ class Vector {
...
         }
 
         // Validate the CVSS v4.0 string vector
-        if (!this.validateStringVector(vector)) {
-            throw new Error("Invalid CVSS v4.0 vector: " + vector);
-        }
+        // if (!this.validateStringVector(vector)) {
```
Comment (skontar) [inline cvss40.js]: Remove commented out code.

File: cvss40.js [hunk 3]
```diff
@@ -880,7 +894,7 @@ class CVSS40 {
             // If the input is a string, create a new Vector object from the string
             this.vector = new Vector(input);
         } else {
-            throw new Error("Invalid input type for CVSS40 constructor. Expected a string or a Vector object.");
+            throw new Error(`Invalid input type for CVSS40 constructor. Expected a string or a Vector object. in ${vector}`);
```
Comment (skontar) [inline cvss40.js]: Full stop in the middle of the sentence.

File: cvss40.test.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    const lineEntries = fileData.split('\n');
+    const scoredVectors = lineEntries.map(vectorScore => {
+      const vectorScorePair = vectorScore.trim().split(' - ');
+      return (vectorScorePair.length !== 2) ? null : { vector: vectorScorePair[0], score: parseFloat(vectorScorePair[1]) };
+    }).filter(Boolean);
+    data[file.name] = scoredVectors;
+    return data;
+  }, {});
+
+  const randomlyPickFrom = (count, array) => array.slice(0, count).sort(() => 0.5 - Math.random());
```
Comment (skontar) [inline cvss40.test.js]: Do we really need to pick random sample? I am not a fan of doing that, non-deterministic tests can be hard to debug.

**Review by skontar — CHANGES_REQUESTED** [review]

File: cvss40.test.js (hunk 4 above)
Comment (superbuggy) [reply cvss40.test.js]: I'm not really a fan either, but had been under the impression this was a requested feature. I was running into max callstack errors with the testing implementation but maybe there's a way I can restructure or force gc and avoid that.

Comment (superbuggy) [reply cvss40.test.js]: We are able to use all data for the tests with my most recent implementation.

File: cvss40.js (hunk 3 above)
Comment (superbuggy) [reply cvss40.js]: Thank you

File: cvss40.js (hunk 2 above)
Comment (superbuggy) [reply cvss40.js]: Done.

File: cvss40.js (hunk 1 above)
Comment (superbuggy) [reply cvss40.js]: Good catch, thank you.

**Review by skontar — APPROVED** [review]

**Comment by pandatix** [comment]
Arbitrary metrics order is invalid according to spec...

Furthermore, the current approach within this SIG for testing is developing an assessment framework so developers can easily handle this part. This PR goes against this effort by bypassing it.
The CVSS v4 calculator serves as a reference for other librarires maintainers, but implementing 500k lines of code just for testing is not acceptable to me.

**Comment by skontar** [comment]
@pandatix OK, I understand that arbitrary metric order is not correct, that makes sense. 

However, I do not understand how adding tests is a problem and "bypassing" any other effort. If this is a reference implementation, why having tests which pass and which can be re-used by others, is wrong? Please elaborate. Also, it is not 500k lines of code, but of test vectors.

I am completely fine removing it from this project, just please give a more detailed rationale. Thanks!

**Comment by pandatix** [comment]
Hey, sorry for the late answer everyone.

Adding tests is not directly "bypassing" the SIG efforts, it is a good thing actually. From an external point of view (like an implementation maintainer/developer) the RedHat implementation is the _de facto_ source of trust. NIST used it for their own calculator, many others accoring to [our analysis](https://github.com/FIRSTdotorg/cvss-working/issues/10).

This PR comes in parallel of this effort: one of the requirements we drew is to provide this test suite in an efficient manner (e.g., a file served by a CDN, a versionned file on GitHub, a GH release asset) and with a limited set of test cases (under 10k if possible) that would cover most cases without needing exhaustivity. Another one is to provide both valid and invalid test cases.
Indeed, we cannot expect an implementation to handle 5GB of data to say that we are confident enough, especially when the implementation is not performant (CI would take minutes up to hours to run at every commit, quickly becoming a burden for maintainers).

To avoid having parallel efforts, I think we should wait for the official FIRST.ORG SIG CVSS test data rather than creating one, propagating it to the community, then having to reach out maintainers to update their work one more time (update source of trust, consumption strategy, data model). We should hop on coordination rather than individual efforts bringing noise to the effort of CVSS v4 adoption.

**Comment by skontar** [comment]
@pandatix I see. So if I understand correctly, the best course of action is to wait for official FIRST.ORG SIG CVSS test vectors, right?

**Comment by pandatix** [comment]
Yes, I think so :smile:

**Pull request merged**